Jeetrohan mpu60xx 1 (BSP-568)

jeetrohan opened the pull request:
# ESP-BSP Pull Request checklist

> Note: For new BSPs create a PR with this [link](https://github.com/espressif/esp-bsp/compare/main...my-branch?quick_pull=1&template=pr_template_bsp.md).
- [ ] Version of modified component bumped
- [ ] CI passing

## Related
- https://github.com/espressif/esp-bsp/issues/317

# Change description

A new mpu60xx library which uses the new I2C driver for reading raw and floating point values; from accelerometer, gyrometer and temperature sensor.
Added functionality to use inbuilt motion detection on mpu60xx using both, polling and interrupts.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-bsp?pullRequest=416) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-bsp?pullRequest=416) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-bsp?pullRequest=416) it.</sub>

tore-espressif commented on components/mpu60xx/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+idf_component_register(SRCS "mpu60xx.c"
+                    INCLUDE_DIRS "include")
```
My initial idea was to release this version as v2.0.0 of the original mpu6050 component. 

This way we won't have 2 driver for the same device...

tore-espressif commented on components/mpu60xx/examples/mpu_motion_detect/CmakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+set(COMPONENTS main)
+include($ENV{IDF_PATH}/tools/cmake/project.cmake)
+project(mpu60xx_motion_detect)
```
Thanks for the example!

jeetrohan replied:
I believe any driver should have at least one example to demonstrate its use.
I'm waiting for some additional component, I'll be providing one more additional example once I receive it. 
And I'm sure it'll be liked even more by the team (I intend to add one more newly released driver from the repository in this one).

tore-espressif replied:
We have some basic guidelines for examples in esp-idf project https://docs.espressif.com/projects/esp-idf/en/latest/esp32/contribute/creating-examples.html

The example overall looks good to me

jeetrohan replied:
Thank you @tore-espressif. 
I'll follow this in next update.

tore-espressif commented on components/mpu60xx/idf_component.yml:
```diff
@@ -0,0 +1,27 @@
+
+description: MPU60xx I2C driver
+ 
+maintainers: 
+  - "Rohan Jeet <jeetrohan92@gmail.com>"
```
This is the most important thing we need to agree on.

If you list yourself as a maintainer, then all bugs and feature request will go your way. You should only add this, if you want to commit yourself to maintenance. It's also possible to list you as 'author' and Espressif will take care of the maintenance.

I can also see that you did not add a license to this new component. All components in this repo must be Apache-2 licensed (that is every one can use for free). If you, as author, require more strict license you will have to publish it in your own repository.

So the bottom line is: if you want to maintain, or don't want to release under Apache-2, or want to release this independently from the original mpu6050, I'd suggest releasing the component on your own.
If all of the above is false, we can review and merge in esp-bsp :)

jeetrohan replied:
> @jeetrohan That's a pretty good job for your first contribution!
I learned the basics from who was considered to be one of the best in my country and I also tried to follow the guidelines in documentation and the video from VP in last year espressif DevCon. 

He (my teacher) strongly advocated opensource.
So ...
1. **Apache 2.0? : Yes**. I want my contribution to be useful to as many people as possible.
Now the only question is how do I do it? Add another file to and make another PR?(This is the only thing I can think of with my limited knowledge of git, as mentioned earlier I'm newbie on git.) 

2.  **release this independently?? :** As I mentioned in another comment its **up to you/the maintenance team**. 
I changed the _init()_  to ensure some flexibility for integrations with rest of the framework. So _**do whatever is suitable, serves the purpose**_.

3. **Author /Maintainer ?** : _Yes. It'd be great if I'm mentioned **just as Author**_ and Espressif as the maintainer takes on all the brunt of resolving the issues (:stuck_out_tongue:).  Although I would like to help resolving the problems even later, whenever possible.

 I believe my answer to Apache2.0 should clarify a lot of things in itself, but if you believe there are unresolved issues, please let me know.

tore-espressif commented on components/mpu60xx/idf_component.yml:
```diff
@@ -0,0 +1,27 @@
+dependencies:
+  ## Required IDF version
+  idf:
+    version: ">=4.1.0"
```
the new I2C driver was added in v5.2
```suggestion
    version: ">=5.2.0"
```

jeetrohan replied:
This was the default from the idf.py and I wasn't mindful even though I thought about looking into it earlier. 
I'd be thankful if you'll update this for me.

tore-espressif replied:
If you're new to ESP-IDF, understanding the various releases can be challenging. The [documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/versions.html) provides an overview if you'd like to learn more.

Generally, we recommend starting with the latest stable release and only upgrading to bugfix releases, unless a specific new feature is required.

However, for driver development,you should specify which ESP-IDF versions your driver supports. The latest I2C driver update is detailed in the "Major Features" section of [v5.2 release](https://github.com/espressif/esp-idf/releases/tag/v5.2). The team is also working on improving how new features and changes are presented.

tore-espressif reviewed (commented):
@jeetrohan That's a pretty good job for your first contribution!

jeetrohan commented:
I just realized the merge here is different from the fork where sync simply updates the branch. :sweat_smile:

jeetrohan commented:
@espzav 
If I'm not wrong the pre-commit failed because of incorrect indentation, which it has fixed (or so it indicates in output).
A rerun should do it.

tore-espressif commented:
> @espzav , @tore-espressif If I'm not wrong the pre-commit failed because of incorrect indentation, which it has fixed (or so it indicates in output). A rerun should do it.

The pre-commit job is always run on latest changes, so there are probably still some formating errors.

If you want to fix them locally, in your esp-bsp folder you can do 
```
pip install pre-commit && pre-commit install
```
To install pre-commit. Then you can run the pre-commit check on you last 11 commits like this
```
pre-commit run --from-ref HEAD~11 --to-ref HEAD
```

tore-espressif commented:
@jeetrohan Thanks again for the PR, I'll take over and fix the remaining things when I'm free

jeetrohan commented:
> > @espzav , @tore-espressif If I'm not wrong the pre-commit failed because of incorrect indentation, which it has fixed (or so it indicates in output). A rerun should do it.
> 
> The pre-commit job is always run on latest changes, so there are probably still some formating errors.
> 
> If you want to fix them locally, in your esp-bsp folder you can do
> 
> ```
> pip install pre-commit && pre-commit install
> ```
> 
> To install pre-commit. Then you can run the pre-commit check on you last 11 commits like this
> 
> ```
> pre-commit run --from-ref HEAD~11 --to-ref HEAD
> ```

@tore-espressif 
Thank you.
I'll do this at the earliest. I'd have done it since the beginning but as I mentioned _it was my first time using the git._ So, I think I'll still need to look more into some documentation to do this properly.
I'll follow these for the next update onward too, so that merging process is not hindered by these simple problems which should be resolved at my end itself.

jeetrohan commented:
> @jeetrohan Thanks again for the PR, I'll take over and fix the remaining things when I'm free

@tore-espressif I totally understand that. I myself had thought to do pre-commit test on branch, but due to some other liabilities and my lack of knowledge (about git itself) I wasn't able to do it ( even though I really want to).
I'll keep you posted, once I'll add more functionality (& this time with pre-commit check. I believe I should've been the one to do it.)
Thanks again.

jeetrohan commented:
hi @tore-espressif & @espzav 
I know its been quite some time since I made the last commit.
There were some irl issues after then, and later I was unwilling to touch again mpu60xx because its quite the subsystem in itself.
I've made a recent commit which is pre-committed and also has the features I promised earlier.
There are still many things I wanted to add.
But I wasn't sure if this PR is still acceptable. 
Please review and let me know if I need to make changes or something else.

jeetrohan commented:
@tore-espressif I think I'll need to update the idf version to >=5.3.
or
I can use i2c_bus component from component registry for backward compatibility.

Which one of the above do you suggest?

jeetrohan commented:
@tore-espressif , @espzav 
**from idf_component.yml in the most recent update:**
```
dependencies:
  idf: ">=5.2"
```

**from the error log :**

```
.Using component placed at /__w/esp-bsp/esp-bsp/components/mpu60xx for dependency mpu60xx(*), specified in /__w/esp-bsp/esp-bsp/components/mpu60xx/examples/mpu_motion_detect/main/idf_component.yml
.HINT: Please check manifest file of the following component(s): main
CMake Error at /opt/esp/idf/tools/cmake/build.cmake:548 (message):
  WARNING: Component "idf" not found

  WARNING: Component "idf" not found

  WARNING: Component "idf" not found

  WARNING: Component "idf" not found

  ERROR: Because project depends on idf (>=5.4.0) which doesn't match any
  versions, version solving failed.
```

And also the the above log is from [build (release-v5.1, 2, 1)](https://github.com/espressif/esp-bsp/actions/runs/15450208889/job/43517321426#logs) 
I am not sure if the run should be performed on idf<5.2.

If I'm wrong let me know.
thanks.

tore-espressif commented:
@jeetrohan thanks for your patience and for all the hard work you’ve put into these updates

Your implementation of the advanced **MPU60xx** features is very valuable for projects that need that extra functionality. At the same time, the goal of **esp-bsp** is to stay focused on a core set of drivers and examples that we can reliably maintain over time. Because the MPU60xx enhancements extend beyond what we normally cover in esp-bsp, we worry that adding them here could make it harder to keep everything tested and up to date.

To make sure your work gets the visibility it deserves—and so that you can continue to iterate on those advanced features without constraints—it would be great to publish this as its own component in the [ESP Registry](https://components.espressif.com/) under your GitHub namespace. That way:

1. **You stay in full control of the code.**  
   You’ll be free to add new features, squash bugs, or make breaking changes as needed, without impacting the esp-bsp maintainers’ stability guarantees.

2. **Users can still pull it in very easily.**  
   Anyone using IDF Component Manager can add your repo to their prject and get those MPU60xx enhancements right away.

3. **esp-bsp remains lean and maintainable.**  
   We can continue focusing on the drivers and examples that serve as a stable, well-tested foundation, while your component lives and evolves alongside it.

The process to register a component in the ESP Registry is covered in the [IDF Component Manager docs](https://docs.espressif.com/projects/idf-component-manager/en/latest/guides/packaging_components.html). You can sign in with your GitHub account, and within a few minutes have it listed. Once it’s published, folks who rely on esp-bsp can still pull in your MPU60xx component as an additional dependency.

I really appreciate your contribution here—this is exactly the kind of work we want to encourage—and I think publishing it as a standalone, registry-managed component is the best path forward so both projects can thrive. Let me know if you need any help with the IDF Component Manager setup!

jeetrohan commented:
@tore-espressif 
Thankyou for your kind words. I'll be closing this PR with this final comment because I believe if there was a way to merge you would have suggested that, not the reasons you have given because
**_This PR was refactored specifically to make maintenance of this driver possible. 
And if I wanted to stay in control I would not have submitted a PR to a official project._**

Assuming 

> MPU60xx enhancements extend beyond what we normally cover in esp-bsp

I'd request you or someone on your team to review :
 _refactor(bh1750): migrate to esp_i2c_driver, add example and update test app (BSP-681)_

It is a simple refactor as it says  and has been pending for over a week. Maybe I can remove the example I added, based off the test app.

And the **Component Registry**
**_Please do not take the following personally, it is not directed at you but a simple observation:_**
The components registered in there are used only if they are from Espressif namespace.
I've made a submission once and probably never will again.


PS:
I would have rather preferred an honest response, older boards are now obsolete and we have moved towards supporting the new ones which do not require these sensors (MPU6050, BH1750, etc..), anymore. 
These are here just for backward compatibility now.
Probably you didn't say this because you are not allowed to divulge the future policies, and I can understand that too.

tore-espressif commented:
I am allowed to say whatever apart from insulting people :D 

The more straightforward answer is: We cannot justify spending time reviewing >1500 line of code from external contributors.
